Feature/mesoscale restart NAM/RAP stats

PerryShafran-NOAA opened the pull request:
<b>Note to developers: You must use this PR template!</b>

## Description of Changes

> Please include a summary of the changes and the related GitHub issue(s). Please also include relevant motivation and context.

Relating to issue #533 , restart for NAM and RAP stats, which is related to Bugzilla 1547.  This PR updates the restart so each file is copied to the restart directory as it is created.  

## Developer Questions and Checklist
* Is this a high priority PR? If so, why and is there a date it needs to be merged by?

No.


* Do you have any planned upcoming annual leave/PTO?

No.


* Are there any changes needed in the times when the jobs are supposed to run/kick-off?

There might be but that will not be tested in this PR and will be considered for a future PR.  
  
- [ ] The code changes follow [NCO's EE2 Standards](https://www.nco.ncep.noaa.gov/idsb/implementation_standards/ImplementationStandards.v11.0.0.pdf).
- [ ] Developer's name is removed throughout the code and have used `${USER}` where necessary throughout the code.
- [ ] References the feature branch for `HOMEevs` are removed from the code.
- [ ] J-Job environment variables, COMIN and COMOUT directories, and output follow what has been [defined](https://docs.google.com/document/d/1JWg_4q80aYmmAoD21GFjp9R9y5-3w7WGM3-0HJk0Pjs/edit#heading=h.7ysbr191vzu4) for EVS.
- [ ] Jobs over 15 minutes in runtime have restart capability.
- [ ] If applicable, changes in the `dev/drivers/scripts` or `dev/modulefiles` have been made in the corresponding `ecf/scripts` and `ecf/defs/evs-nco.def`? 
- [ ] Jobs contain the appropriate file checking and don't run METplus for any missing data.
- [ ] Code is using METplus wrappers structure and not calling MET executables directly.
- [ ] Log is free of any ERRORs or WARNINGs.

## Testing Instructions

> Please include testing instructions for the PR assignee. Include all relevant input datasets needed to run the tests.

1) Clone the repository.
2) Checkout the branch feature/mesoscale_restart.
3) Link the fix directory.
4) Go to the directory dev/drivers/scripts/stats/mesoscale
5) In the files jevs_mesoscale_nam_stats.sh and jevs_mesoscale_rap_stats.sh, add in the line:  
`export EVSINspcotlk=/lfs/h2/emc/vpppg/noscrub/emc.vpppg/evs/v2.0/prep/cam`
6) Run each of these scripts, using the options `qsub -v vhr=07`
  a) Run as normal to get a baseline for the final size of the stats file.  Save this file to compare to the final file for restart.  
  b) Set the wallclock to 15 minutes to get an interrupted run.  You know the run is interrupted when you see the wallclock exceeded message and some SIGTERM error messages, and there is no final stats file written.  
  c) Set the wallclock to the standard time (1 hr for NAM, 1 hr 30 min for RAP) and run again, without clearing the small stats directory.  This will get you the final stats file to compare to the baseline from step 6a).

PerryShafran-NOAA committed:
Adding the restart items to the ush directory

PerryShafran-NOAA committed:
Updating evs_config files with the restart settings

PerryShafran-NOAA committed:
More ush and ex-script updates

PerryShafran-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/mesoscale_restart

Commit:
Adding in new stuff for restart

PerryShafran-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/mesoscale_restart

PerryShafran-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/mesoscale_restart

PerryShafran-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/mesoscale_restart

PerryShafran-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/mesoscale_restart

Commit:
More changes to correct the restart with mpmd

Commit:
Updating for rap and completing restart fixes

PerryShafran-NOAA commented on ush/mesoscale/mesoscale_production_restart.py:
```diff
@@ -20,20 +23,33 @@
             )
 
+
 print("END: "+os.path.basename(__file__))
```
@malloryprow Before you review, I just noticed the differences in this particular file seem to indicate that I copied over the pre-mpmd version of this script that doesn't have the working directory stuff for mpmd.  I'm not terribly sure if it makes any difference, but I wanted to do a quickie test using the updated restart file to know that we have the correct restart script with mpmd.  Hold a quick sec while I make this test.

malloryprow replied:
Let me know how the test goes!

PerryShafran-NOAA replied:
It didn't matter which version of this py script was used, the final result was the same size of final stat file compared to the stat file without restart.  I think that this is the more correct version since I think changes were made in this file in a previous recent PR.

PerryShafran-NOAA replied:
@malloryprow I need to make further updates to this file.  Turns out that the old one was actually correct.

PerryShafran-NOAA replied:
Please continue to hold this PR.

PerryShafran-NOAA replied:
Updated to the new file, which is the correct one.  We can resolve this particular conversation.

malloryprow commented:
A few things:

1. I noticed this in #647. The completed files for restart are being written directly to COMOUT. It should be written to DATA and then copied to COMOUT if `SENDCOM=YES`.

2. Since the function `copy_data_to_restart` is copying to COMOUT. There should be a check for `SENDCOM=YES`.

3. There are many lines in ush/mesoscale/mesoscale_stats_grid2obs_create_job_script.py that start with `#python -c` or `#{metplus_launcher}`. If this is being written to the job file as a commented out line, is it needed at all?

PerryShafran-NOAA commented:
I'd like @MarcelCaron-NOAA to comment on these items.  He had already addressed 3 elsewhere, but it's worth repeating here, but I don't recall the reason for that, but there was a reason.  

I also thought that files were generated in the working directories and then copied to COMOUT, again the impression that Marcel gave me.  Is this not the case?

malloryprow commented:
For 1., I see export `RESTART_DIR="${COMOUTsmall}/restart/c${vhr}"` in parm/evs_config/mesoscale/config.evs.prod.stats.mesoscale.atmos.grid2obs.[nam][rap]. Then in ush/mesoscale/mesoscale_stats_grid2obs_create_job_script.py one line 38 this get read in by `RESTART_DIR = os.environ['RESTART_DIR']`.  Then on lines 413-418, for example, is `mesoscale_util.mark_job_completed("+ f"\"{os.path.join(RESTART_DIR, COMPLETED_JOBS_FILE)}`.  The same sort of logic and be followed for 2. but look at `copy_data_to_restart`.

PerryShafran-NOAA commented:
@malloryprow I'm looking through one of my .o files, and I can only see instances of data being copied from the working directory DATA to the COMOUT restart directory.  

However, I do believe that you may be thinking about this?

```
26 + python -c 'import mesoscale_util; mesoscale_util.mark_job_completed("/lfs/h2/emc/vpppg/noscrub/perry.shafran/evs/v2.0/stats/
mesoscale/atmos.20250113/nam/grid2obs/restart/c07/completed_jobs.txt", "job36", job_type="generate")'
```

So basically the completed_jobs.txt file is actually the file that's being directly written in COMOUT, but data itself is being generated in DATA and copied to COMOUT.  Just this file is the one that is being written directly.  Is this file the issue?

PerryShafran-NOAA commented:
OK, your comment confirms that it is the completed_jobs.txt file that's at issue. I'll need to figure out how this file can be written in DATA and then copied over to COMOUT at the end of the run.

malloryprow commented:
Okay about the restart; the code looks like it is doing that but it must not be! Is the code checking that `SENDCOM=YES` before copying to COMOUT?

And yup, the completed file can't be written directly to COMOUT just like all other files.

PerryShafran-NOAA commented:
As far as I know there aren't any checks for SENDCOM anywhere in the python scripts for restart.  

So there are two issues here to deal with:

1) The completed_jobs.txt file is being written directly to COMOUT and it shouldn't be.
2) There are no checks for SENDCOM.  

But all data itself (obs files, stat files, etc) are being written in DATA and then copied over to COMOUT.  Only the completed_jobs.txt file is being written directly in COMOUT.

malloryprow commented:
Yes, sounds good! Let me know if you need any help!

PerryShafran-NOAA commented:
I think I may need help but I'll contact you off this PR thread and not muddle this thread with coding assistance.

PerryShafran-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/mesoscale_restart

MarcelCaron-NOAA commented:
> I'd like @MarcelCaron-NOAA to comment on these items. He had already addressed 3 elsewhere, but it's worth repeating here, but I don't recall the reason for that, but there was a reason.

The commented-out commands help with debugging. If only "generate/job2" completed successfully, we could skip writing it in a restart run (though that would require code changes). However, during development and debugging, it's useful to see on the card what was skipped and what the environment would have been, for example when trying to track down missing data or test-run the job on its own. This way, the successfully completed job card reflects what would have run, and, when submitted during restart, completes quickly.

MarcelCaron-NOAA commented:
> * The completed_jobs.txt file is being written directly to COMOUT and it shouldn't be.
> * There are no checks for SENDCOM.
* I haven't looked deeply but this seems to be the case.  The fix would likely be [here](https://github.com/NOAA-EMC/EVS/blob/7533c07ea6b199a01cc5981e06cd3788e04a1845/ush/cam/cam_util.py#L329C1-L334C37) (for cam, same function in the equivalent file for mesoscale).  To fix, we could maybe write `SENDCOM` to each job card, feed in the `DATA` and `SENDCOM` values to the above function, [grab the filename from "`completed_jobs_file`" path](https://stackoverflow.com/questions/8384737/extract-file-name-from-path-no-matter-what-the-os-path-format), write the file to `DATA/filename` rather than to `completed_jobs_file`, then check `SENDCOM` and copy `DATA/filename` to `completed_jobs_file`.
* Yup, cam also copies to comout without checking if `SENDCOM=YES`.  The fix may be ~simple if we export "`SENDCOM`" in each job card, then feed that value into "copy_data_to_restart()", checking the value maybe before [lines 718-720](https://github.com/NOAA-EMC/EVS/blob/7533c07ea6b199a01cc5981e06cd3788e04a1845/ush/cam/cam_util.py#L718C1-L720C18) in ush/cam/cam_util.py and the equivalent lines in ush/mesoscale/mesoscale_util.py.

MarcelCaron-NOAA commented:
> > * The completed_jobs.txt file is being written directly to COMOUT and it shouldn't be.
> > * There are no checks for SENDCOM.
> 
> * I haven't looked deeply but this seems to be the case.  The fix would likely be [here](https://github.com/NOAA-EMC/EVS/blob/7533c07ea6b199a01cc5981e06cd3788e04a1845/ush/cam/cam_util.py#L329C1-L334C37) (for cam, same function in the equivalent file for mesoscale).  To fix, we could maybe write `SENDCOM` to each job card, feed in the `DATA` and `SENDCOM` values to the above function, [grab the filename from "`completed_jobs_file`" path](https://stackoverflow.com/questions/8384737/extract-file-name-from-path-no-matter-what-the-os-path-format), write the file to `DATA/filename` rather than to `completed_jobs_file`, then check `SENDCOM` and copy `DATA/filename` to `completed_jobs_file`.
> * Yup, cam also copies to comout without checking if `SENDCOM=YES`.  The fix may be ~simple if we export "`SENDCOM`" in each job card, then feed that value into "copy_data_to_restart()", checking the value maybe before [lines 718-720](https://github.com/NOAA-EMC/EVS/blob/7533c07ea6b199a01cc5981e06cd3788e04a1845/ush/cam/cam_util.py#L718C1-L720C18) in ush/cam/cam_util.py and the equivalent lines in ush/mesoscale/mesoscale_util.py.

On second thought, `SENDCOM` can be checked in the job card itself ("if `SENDCOM`, then copy to restart")

Commit:
Keeping current develop version of meoscale_production_restart.py

PerryShafran-NOAA commented:
> * I haven't looked deeply but this seems to be the case.  The fix would likely be [here](https://github.com/NOAA-EMC/EVS/blob/7533c07ea6b199a01cc5981e06cd3788e04a1845/ush/cam/cam_util.py#L329C1-L334C37) (for cam, same function in the equivalent file for mesoscale).  To fix, we could maybe write `SENDCOM` to each job card, feed in the `DATA` and `SENDCOM` values to the above function, [grab the filename from "`completed_jobs_file`" path](https://stackoverflow.com/questions/8384737/extract-file-name-from-path-no-matter-what-the-os-path-format), write the file to `DATA/filename` rather than to `completed_jobs_file`, then check `SENDCOM` and copy `DATA/filename` to `completed_jobs_file`.

@MarcelCaron-NOAA Just wanting to confirm, the link "here" for the fix would be the fix to ensure that the completed_jobs.txt file is written in DATA and then copied to COMOUT?

PerryShafran-NOAA commented:
@MarcelCaron-NOAA It should be noted that the block of code highlighted in your link to the fix already exists in my mesoscale_util.py file.  See line 382 here:

https://github.com/PerryShafran-NOAA/EVS/blob/feature/mesoscale_restart/ush/mesoscale/mesoscale_util.py

MarcelCaron-NOAA commented:
> > * I haven't looked deeply but this seems to be the case.  The fix would likely be [here](https://github.com/NOAA-EMC/EVS/blob/7533c07ea6b199a01cc5981e06cd3788e04a1845/ush/cam/cam_util.py#L329C1-L334C37) (for cam, same function in the equivalent file for mesoscale).  To fix, we could maybe write `SENDCOM` to each job card, feed in the `DATA` and `SENDCOM` values to the above function, [grab the filename from "`completed_jobs_file`" path](https://stackoverflow.com/questions/8384737/extract-file-name-from-path-no-matter-what-the-os-path-format), write the file to `DATA/filename` rather than to `completed_jobs_file`, then check `SENDCOM` and copy `DATA/filename` to `completed_jobs_file`.
> 
> @MarcelCaron-NOAA Just wanting to confirm, the link "here" for the fix would be the fix to ensure that the completed_jobs.txt file is written in DATA and then copied to COMOUT?

@PerryShafran-NOAA Yes, the fix to write in DATA then copy to COMOUT

PerryShafran-NOAA commented:
@malloryprow @MarcelCaron-NOAA I think the way to solve this would be to do the following, and I think it would be in mesoscale_stats_grid2obs_create_job_script.py is to do the following:

In each case where we see something like:

```
                  job_cmd_list_iterative.append(
                       "python -c "
                       + f"'import mesoscale_util; mesoscale_util.mark_job_completed("
                       + f"\"{os.path.join(RESTART_DIR, COMPLETED_JOBS_FILE)}\", "
                       + f"\"job{njob}\", job_type=\"{job_type}\")'"
```

Which is where the completed jobs file is currently written to the restart directory RESTART_DIR, we need to change RESTART_DIR to something else, maybe a COMPLETED_JOBS_DIR where we can define in the parm file and read in to this script.  Then after this line, we can add a copy_data_to_restart line that starts like this:

```
                   job_cmd_list_iterative.append(
                       f'python -c '
                       + '\"import mesoscale_util as cutil; cutil.copy_data_to_restart('
                       + '\\\"${DATA}\\\", \\\"${RESTART_DIR}\\\", '

```

The additional copy_data_to_restart would then copy this completed jobs file each time to the RESTART_DIR directory, so it continues to grow in COMOUT like it does now.  It would have to be done every time a job is completed, as the file contains critical information as to when a job is completed which would be needed in a restart case.  

The question I have is I'm not to sure how to formulate the copy_data_to_restart command for the completed jobs file.  If I could have some assistance there from either of you, that would be great.

MarcelCaron-NOAA commented:
> @MarcelCaron-NOAA It should be noted that the block of code highlighted in your link to the fix already exists in my mesoscale_util.py file. See line 382 here:
> 
> https://github.com/PerryShafran-NOAA/EVS/blob/feature/mesoscale_restart/ush/mesoscale/mesoscale_util.py

@PerryShafran-NOAA yes the mesoscale bit of code would be the relevant bit for the fix to mesoscale.  I only linked cam because it's not currently being worked on and hasn't been modified from develop!

PerryShafran-NOAA commented:
> > @MarcelCaron-NOAA It should be noted that the block of code highlighted in your link to the fix already exists in my mesoscale_util.py file. See line 382 here:
> > https://github.com/PerryShafran-NOAA/EVS/blob/feature/mesoscale_restart/ush/mesoscale/mesoscale_util.py
> 
> @PerryShafran-NOAA yes the mesoscale bit of code would be the relevant bit for the fix to mesoscale. I only linked cam because it's not currently being worked on and hasn't been modified from develop!

I'm saying that your fix from the cam script already exists in the mesoscale script, if I'm looking at my own code correctly.

PerryShafran-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/mesoscale_restart

PerryShafran-NOAA commented:
To clarify, this block of code is what you cite as the fix to write the completed jobs file in DATA and then copy to restart:

```
def mark_job_completed(completed_jobs_file, job_name, job_type=""):
    with open(completed_jobs_file, 'a') as f:
        if job_type:
            f.write(job_type + "_" + job_name + "\n")
        else:
            f.write(job_name + "\n")
```

This block of code is already in the mesoscale_util.py script.  And as far as I can tell, the file is still being written directly to COMOUT even with this block of code in there.

Mallory had identified the issue as stemming from the block of code in mesoscale_stats_grid2obs_create_job_script.py that writes directly to COMOUT, which is this:

```
        job_cmd_list.append(
            "python -c "
            + f"'import mesoscale_util; mesoscale_util.mark_job_completed("
            + f"\"{os.path.join(RESTART_DIR, COMPLETED_JOBS_FILE)}\", "
            + f"\"job{njob}\", job_type=\"{job_type}\")'"
```

This needs to be modified I think.

malloryprow commented:
The second block of code is calling the first block of code. Marcel is saying the the first block needs to be modified.

PerryShafran-NOAA commented:
OK, so the block of code, the def_mark_job_completed block, is what needs to be modified from what it is now?

malloryprow commented:
Yes, I believe so.

PerryShafran-NOAA commented:
Oh, OK, that makes this more clear.  Though I'm not entirely sure how this is fixed, but it sounds like Marcel is developing the fix in cam.

Can we start to run the review now, though?  We could get the restart working and merged, and then potentially add the remaining issues to the Fixes and Additions list for a future PR.  I'm also OK with taking the time to get these fixes in now so we get it done and don't have to worry about it later.

malloryprow commented:
I think I'd like to see this fix in this PR then we can say that restart for grid2obs for mesoscale is complete rather than waiting on another PR that addresses issues with the restart.

PerryShafran-NOAA commented:
OK, fair enough.  

@MarcelCaron-NOAA I think you said you're working on the fix for CAM?

MarcelCaron-NOAA commented:
@PerryShafran-NOAA No I am not currently working on this fix, but it does need to be added to cam.  I don't think it will be a complicated fix, but I'm not sure how quickly I'll be able to get to it.

PerryShafran-NOAA commented:
Understood.  Perhaps I can work on this with @malloryprow to get a fix in this code.

Commit:
Restoring the updated mesoscale_production_restart.py file

PerryShafran-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/mesoscale_restart

Commit:
Updating for completed_jobs.txt to copy from DATA to COMOUT

Commit:
Adding and correcting the if existence checks of completed_jobs.txt copy to COMOUT

Commit:
Adding SENDCOM to the create job script

malloryprow commented:
From the commit is looks like the files will now be written to directories `${COMOUTfinal}/precip_mean24/ccpa.${VDATE}` and `${COMOUTfinal}/precip_mean24/href.YYYYmmdd` where YYYYmmdd corresponds to some init date.

Is there a reason they all cannot just be written to `${COMOUTfinal}/precip_mean24`? I don't think NCO is going to like another directory with a date within COMOUTfinal.

Commit:
Updating rap conf file

PerryShafran-NOAA commented:
@malloryprow I have successfully addressed the issues from above:

1) The completed jobs file is now being written in DATA, the working directory, and is sent to COMOUT instead of being written directly to COMOUT.
2) There are SENDCOM checks in place for both the data and also the completed jobs file.  

I think we can now re-start testing for this PR.

malloryprow commented:
COMOUT is /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr652/evs/v2.0/stats/mesoscale.

### 1. jevs_mesoscale_nam_grid2obs_stats.sh (vhr=07) - Full Test
**Log File:** /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr652/evs/v2.0/stats/mesoscale/jevs_mesoscale_nam_grid2obs_stats_00.o178180522
**DATA:** /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_mesoscale_nam_grid2obs_stats_00.178180522.cbqs01

### 2. jevs_mesoscale_rap_grid2obs_stats.sh (vhr=07) - Full Test
**Log File:** /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr652/evs/v2.0/stats/mesoscale/jevs_mesoscale_rap_grid2obs_stats_00.o178180768
**DATA:** /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_mesoscale_rap_grid2obs_stats_00.178180768.cbqs01

PerryShafran-NOAA commented:
@malloryprow Did you pull in all the changes?  I don't see a restart directory in the atmos.20250126/nam/grid2obs or the atmos.20250126/rap/grid2obs directories.

malloryprow commented:
Believe so. The last commit in my log for my branch is "Updating rap conf file".

I see /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr652/evs/v2.0/stats/mesoscale/atmos.20250126/nam/grid2obs and /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr652/evs/v2.0/stats/mesoscale/atmos.20250126/rap/grid2obs.

PerryShafran-NOAA commented:
Lemme see if I included all the updates in my commits.

Commit:
Updating mesoscale_stats_grid2obs_create_job_script.py

PerryShafran-NOAA commented:
Can you clear the directory and run everything again?  I have an update.  I have NO clue why I didn't update this one.

malloryprow commented:
My branch has the latest commit in the log "Updating mesoscale_stats_grid2obs_create_job_script.py"

COMOUT is /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr652/evs/v2.0/stats/mesoscale.

### 1. jevs_mesoscale_nam_grid2obs_stats.sh (vhr=07) - Full Test
**Log File:** /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr652/evs/v2.0/stats/mesoscale/jevs_mesoscale_nam_grid2obs_stats_00.o178184142
**DATA:** /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_mesoscale_nam_grid2obs_stats_00.178184142.cbqs01

### 2. jevs_mesoscale_rap_grid2obs_stats.sh (vhr=07) - Full Test
**Log File:** /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr652/evs/v2.0/stats/mesoscale/jevs_mesoscale_rap_grid2obs_stats_00.o178184144
**DATA:** /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_mesoscale_rap_grid2obs_stats_00.178184144.cbqs01

PerryShafran-NOAA commented:
Whew, I see the restart directory in your newest run.

malloryprow commented:
I'm seeing some thing in /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr652/evs/v2.0/stats/mesoscale/jevs_mesoscale_nam_grid2obs_stats_00.o178184142

1. forrtl: error (78): process killed (SIGTERM)
2. CFP RANK  19    CFP TASK NUMBER: 0020 FAILED.     USER COMMAND: /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_mesoscale_nam_grid2obs_stats_00.178184142.cbqs01/grid2obs/stats/METplus_job_scripts/reformat/job20
3. cp: cannot create regular file '/lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr652/evs/v2.0/stats/mesoscale/atmos.20250126/nam/grid2obs/restart/c07/completed_jobs.txt': File exists
4. nid001101.cactus.wcoss2.ncep.noaa.gov: rank 246 exited with code 1

PerryShafran-NOAA commented:
@malloryprow I am not entirely sure what the issue is here.  I do not see any of the above errors in the script that I just ran.  Let me see if there are updates that I somehow didn't put into the repo, but I'm not seeing these errors.  

My latest .o file is here:  /lfs/h2/emc/vpppg/noscrub/perry.shafran/EVS_mesoscale_restart/EVS/dev/drivers/scripts/stats/mesoscale/jevs_mesoscale_nam_grid2obs_stats_00.o178183762

malloryprow commented:
I can try to run a clean, full test again if you want.

PerryShafran-NOAA commented:
I don't see the .o file in the location where you noted.  Is it moved somewhere else?  

I'd like to check your code base to see if I'm missing any updates.

malloryprow commented:
I cleaned out everything and trying again.

COMOUT is /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr652/evs/v2.0/stats/mesoscale.

### 1. jevs_mesoscale_nam_grid2obs_stats.sh (vhr=07) - Full Test
**Log File:** /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr652/EVS/dev/drivers/scripts/stats/mesoscale/jevs_mesoscale_nam_grid2obs_stats_00.o178186535
**DATA:** /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_mesoscale_nam_grid2obs_stats_00.178186535.cbqs01

### 2. jevs_mesoscale_rap_grid2obs_stats.sh (vhr=07) - Full Test
**Log File:** /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr652/EVS/dev/drivers/scripts/stats/mesoscale/jevs_mesoscale_rap_grid2obs_stats_00.o178186537
**DATA:** /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_mesoscale_rap_grid2obs_stats_00.178186537.cbqs01

PerryShafran-NOAA commented:
OK, your codebase matches mine so you have all the updates.  I'm not why you got those errors.

Yes, clean out all atmos.20250126 directories and run again.

PerryShafran-NOAA commented:
OK, all checks out with the NAM and RAP.  There are no errors, the restart directory is there under atmos.20250126/nam/grid2obs/restart, and atmos.20250126/rap/grid2obs/restart, with the completed_jobs.txt file there and it looks like it should.  The NAM stats file is a bit different than emc.vpppg, but that is because of the timing issue we had discussed before, so it's OK.

Not sure what happened before, but all looks good now.

PerryShafran-NOAA commented:
Oh wait, SIGTERM in the RAP output.  Let me take a quick look-see.

PerryShafran-NOAA commented:
OK, I am seeing this in the RAP output:

```
cp: cannot create regular file '/lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr652/evs/v2.0/stats/mesoscale/atmos.20250126/rap/grid2obs/restart/c07/completed_jobs.txt': File exists
CFP RANK  27    CFP TASK NUMBER: 0028 FAILED.     USER COMMAND: /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_mesoscale_rap_grid2obs_stats_00.178186537.cbqs01/grid2obs/stats/METplus_job_scripts/reformat/job28
```

Same as you saw before with the NAM.

Could it be that errors occur when two processes are trying to write to the completed_jobs.txt at the same time?

malloryprow commented:
It is possible. Two jobs might be trying copy the file at the same time.

PerryShafran-NOAA commented:
I wonder if it would help if I used the -f option:

```
       -f, --force
              if  an  existing  destination  file cannot be opened, remove it and try again (this option is ignored
              when the -n option is also used)
```

Commit:
Adding -f option to the cp of completed jobs file

malloryprow commented:
It may. Could there possibility of a completed job potentially being left off this list? The "remove it" in that help statement is what makes me think that could be possible? Looking at it now, it is concerning all the jobs are writing to the same completed_jobs.txt. NCO doesn't want use having MPMD jobs writing to the same directory so I can't imagine they would like them all writing to the same file.

There was a similar that happened in operations global_det atmos stats last summer.The jobs were all trying to write to the same file and had the same `cp: cannot create regular file...: File exists`. All the grid2grid jobs were trying to copy to the same file. I had to take the CCPA file they were trying to write to and move it to the prep step. That fix doesn't apply here really but just wanted to share.

PerryShafran-NOAA commented:
@malloryprow I just tested with the -f option and it seems to be OK.  

In thinking whether a completed job would be left off the list, I don't think so.  The completed_jobs.txt file in the working directory DATA keeps growing when jobs are completed, and if, perhaps, in one instance, a certain completed job didn't get written to the COMOUT version, it would just be included with the next cp since the version in the working directory would not miss a file completion.  Does that make sense?

PerryShafran-NOAA commented:
Try picking up the latest update, where I include -f, and run some clean jobs once again.  If they pass, we should proceed.

malloryprow commented:
I cleaned out everything and trying again with last commit "Adding -f option to the cp of completed jobs file".

COMOUT is /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr652/evs/v2.0/stats/mesoscale.

### 1. jevs_mesoscale_nam_grid2obs_stats.sh (vhr=07) - Full Test
**Log File:** /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr652/EVS/dev/drivers/scripts/stats/mesoscale/jevs_mesoscale_nam_grid2obs_stats_00.o178196383
**DATA:** /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_mesoscale_nam_grid2obs_stats_00.178196383.cbqs01

### 2. jevs_mesoscale_rap_grid2obs_stats.sh (vhr=07) - Full Test
**Log File:** /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr652/EVS/dev/drivers/scripts/stats/mesoscale/jevs_mesoscale_rap_grid2obs_stats_00.o178196384
**DATA:** /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_mesoscale_rap_grid2obs_stats_00.178196384.cbqs01

PerryShafran-NOAA commented:
All now looks good for both NAM and RAP stats, for the full run.  No errors or SIGTERMs, output stat files are of expected size.

malloryprow commented:
Full test directories moved under /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr652/evs/v2.0/stats/mesoscale/full_test, so COMOUT for restart tests is /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr652/evs/v2.0/stats/mesoscale.

### 1. jevs_mesoscale_nam_grid2obs_stats.sh (vhr=07) - Restart Test
**Full Test Run Runtime**: 20 minutes
**Interrupted Runtime**: 10 minutes
**Interrupted Log File**: /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr652/EVS/dev/drivers/scripts/stats/mesoscale/jevs_mesoscale_nam_grid2obs_stats_00.o178199577
**Interrupted DATA**: /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_mesoscale_nam_grid2obs_stats_00.178199577.cbqs01
**Restart Runtime**: 20 minutes
**Restart Log File**: /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr652/EVS/dev/drivers/scripts/stats/mesoscale/jevs_mesoscale_nam_grid2obs_stats_00.o178200454
**Restart DATA:** /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_mesoscale_nam_grid2obs_stats_00.178200454.cbqs01

### 2. jevs_mesoscale_rap_grid2obs_stats.sh (vhr=07) - Restart Test
**Full Test Run Runtime**: 26 minutes
**Interrupted Runtime**: 10 minutes
**Interrupted Log File**:  /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr652/EVS/dev/drivers/scripts/stats/mesoscale/jevs_mesoscale_rap_grid2obs_stats_00.o178199582
**Interrupted DATA**: /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_mesoscale_rap_grid2obs_stats_00.178199582.cbqs01
**Restart Runtime**: 22 minutes
**Restart Log File**: /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr652/EVS/dev/drivers/scripts/stats/mesoscale/jevs_mesoscale_rap_grid2obs_stats_00.o178200466
**Restart DATA:** /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_mesoscale_rap_grid2obs_stats_00.178200466.cbqs01

PerryShafran-NOAA commented:
@malloryprow Can we run these again, but only the interrupted job for now?  The interruption came while the child directories were still being created, or job scripts are still being created, thus I think we need to increase the run time to 15 minutes, as no data had been generated yet.  I want to see if any data is being created and a completed_jobs.txt file created before we go to the restart run.

PerryShafran-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/mesoscale_restart

malloryprow commented:
Full test directories moved under /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr652/evs/v2.0/stats/mesoscale/full_test, so COMOUT for restart tests is /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr652/evs/v2.0/stats/mesoscale.

### 1. jevs_mesoscale_nam_grid2obs_stats.sh (vhr=07) - Restart Test
**Full Test Run Runtime**: 20 minutes
**Interrupted Runtime**: 15 minutes
**Interrupted Log File**: /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr652/EVS/dev/drivers/scripts/stats/mesoscale/jevs_mesoscale_nam_grid2obs_stats_00.o178206761
**Interrupted DATA**: /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_mesoscale_nam_grid2obs_stats_00.178206761.cbqs01
**Restart Runtime**: 19 minutes
**Restart Log File**: /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr652/EVS/dev/drivers/scripts/stats/mesoscale/jevs_mesoscale_nam_grid2obs_stats_00.o178207783
**Restart DATA:** /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_mesoscale_nam_grid2obs_stats_00.178207783.cbqs01

### 2. jevs_mesoscale_rap_grid2obs_stats.sh (vhr=07) - Restart Test
**Full Test Run Runtime**: 26 minutes
**Interrupted Runtime**: 15 minutes
**Interrupted Log File**:  /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr652/EVS/dev/drivers/scripts/stats/mesoscale/jevs_mesoscale_rap_grid2obs_stats_00.o178206765
**Interrupted DATA**: /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_mesoscale_rap_grid2obs_stats_00.178206765.cbqs01
**Restart Runtime**: 26 minutes
**Restart Log File**: /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr652/EVS/dev/drivers/scripts/stats/mesoscale/jevs_mesoscale_rap_grid2obs_stats_00.o178207786
**Restart DATA:** /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_mesoscale_rap_grid2obs_stats_00.178207786.cbqs01

PerryShafran-NOAA commented:
@malloryprow Sigh, there is another ERROR.  I feel it is easily solvable, but I'd need to wrap my head as to how to figure it out in the code.

In the restart run, I am seeing errors relating to ptype, for example:

`WARNING: Missing input files (/lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_mesoscale_nam_grid2obs_stats_00.178207783.cbqs01/grid2obs/METplus_output/workdirs/generate/job361/metar/regrid_data_plane/nam.20250126/regrid_data_plane_nam_t00z_metar_conusp_job361_fhr00.nc) cannot make output file /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_mesoscale_nam_grid2obs_stats_00.178207783.cbqs01/grid2obs/data/nam/merged_ptype/merged_ptype_metar_conusp_job361_init2025012600_fhr000.nc`

and 

`01/27 20:33:43.180 metplus.9a8b2de9 (command_builder.py:264) ERROR: (command_builder.py:619) Could not find FCST_INPUT file /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_mesoscale_nam_grid2obs_stats_00.178207783.cbqs01/grid2obs/data/nam/merged_ptype/merged_ptype_metar_conusp_job361_init2025012600_fhr000.nc using template /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_mesoscale_nam_grid2obs_stats_00.178207783.cbqs01/grid2obs/data/nam/merged_ptype/merged_ptype_metar_conusp_job361_init{init?fmt=%Y%m%d%H}_fhr{lead?fmt=%3H}.nc`

No such errors occur in the full uninterrupted run.  

Now here is what I'm thinking.  The ptype obs were already generated and the restart brought in the stat files from the restart directory.  However, somehow, the ptype processes did not know that they did not need to run ptype, and so thus there are missing files in workdirs that would ordinarily be there.  If that makes any sense.  

FWIW, the final stat files for the restart job are identical to the final stat files for the full uninterrupted job - because the ptype stat files were already generated in the interrupted run so they were not needed to be run again.  

Tagging @MarcelCaron-NOAA for some insight.  There is a missing piece of restart somewhere that should fix this.

MarcelCaron-NOAA commented:
@PerryShafran-NOAA It looks like the needed `regrid_data_plane` output is missing, which I think is needed in order to restart a "`generate`" job that, say, produced the merged ptype file ("`merged_ptype_metar_...nc`") but didn't then finish the PointStat command before the interruption.

Do you know if this `regrid_data_plane` file is produced in the same "`generate`" job (`job361`)?

- If so, there should be another input file that is missing, since this `regrid_data_plane` file wasn't created
- If not, then the `regrid_data_plane` file should exist in the main directory: `/lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_mesoscale_nam_grid2obs_stats_00.178207783.cbqs01/grid2obs/METplus_output/metar/regrid_data_plane/nam.20250126`

... instead of the working directory for "`generate`" job 361.  Maybe METplus is looking in the wrong directory?

These are my initial thoughts not having looked at the log file for this test.

PerryShafran-NOAA commented:
@malloryprow @MarcelCaron-NOAA Here is what is found in job361:

```
    for FHR in `seq ${FHR_START} ${FHR_INCR} ${FHR_END}`; do
        export FHR=$(printf "%02d" $FHR)
        #jobs were restarted, and the following has already run successfully
        #run_metplus.py -c /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr652/EVS/parm/metplus_config/machine.conf -c /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr652/EVS/parm/metplus_config/stats/mesoscale/grid2obs/RegridDataPlane_fcstMESOSCALE_PTYPE.conf
        #python -c "import mesoscale_util as cutil; cutil.copy_data_to_restart(\"${DATA}\", \"${RESTART_DIR}\", njob=\"361\", verif_case=\"${VERIF_CASE}\", verif_type=\"${VERIF_TYPE}\", vx_mask=\"${NEST}\", met_tool=\"regrid_data_plane\", vdate=\"${VDATE}\", vhour=\"${VHOUR}\", fhr_start=\"${FHR_START}\", fhr_end=\"${FHR_END}\", fhr_incr=\"${FHR_INCR}\", model=\"${MODELNAME}\", )"
        #python /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr652/EVS/ush/mesoscale/mesoscale_stats_grid2obs_create_merged_ptype.py
        #python -c "import mesoscale_util as cutil; cutil.copy_data_to_restart(\"${DATA}\", \"${RESTART_DIR}\", njob=\"361\", verif_case=\"${VERIF_CASE}\", verif_type=\"${VERIF_TYPE}\", vx_mask=\"${NEST}\", met_tool=\"merged_ptype\", vdate=\"${VDATE}\", vhour=\"${VHOUR}\", fhr_start=\"${FHR_START}\", fhr_end=\"${FHR_END}\", fhr_incr=\"${FHR_INCR}\", model=\"${MODELNAME}\", )"
        #run_metplus.py -c /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr652/EVS/parm/metplus_config/machine.conf -c /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr652/EVS/parm/metplus_config/stats/mesoscale/grid2obs/PointStat_fcstMESOSCALE_obsMETAR_PTYPE.conf
        #python -c "import mesoscale_util as cutil; cutil.copy_data_to_restart(\"${DATA}\", \"${RESTART_DIR}\", njob=\"361\", verif_case=\"${VERIF_CASE}\", verif_type=\"${VERIF_TYPE}\", vx_mask=\"${NEST}\", met_tool=\"point_stat\", vdate=\"${VDATE}\", vhour=\"${VHOUR}\", fhr_start=\"${FHR_START}\", fhr_end=\"${FHR_END}\", fhr_incr=\"${FHR_INCR}\", model=\"${MODELNAME}\", var_name=\"${VAR_NAME}\")"
        python /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr652/EVS/ush/mesoscale/mesoscale_stats_grid2obs_create_merged_ptype.py
        python -c "import mesoscale_util as cutil; cutil.copy_data_to_restart(\"${DATA}\", \"${RESTART_DIR}\", njob=\"361\", verif_case=\"${VERIF_CASE}\", verif_type=\"${VERIF_TYPE}\", vx_mask=\"${NEST}\", met_tool=\"merged_ptype\", vdate=\"${VDATE}\", vhour=\"${VHOUR}\", fhr_start=\"${FHR_START}\", fhr_end=\"${FHR_END}\", fhr_incr=\"${FHR_INCR}\", model=\"${MODELNAME}\", )"
        run_metplus.py -c /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr652/EVS/parm/metplus_config/machine.conf -c /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr652/EVS/parm/metplus_config/stats/mesoscale/grid2obs/PointStat_fcstMESOSCALE_obsMETAR_PTYPE.conf
        python -c "import mesoscale_util as cutil; cutil.copy_data_to_restart(\"${DATA}\", \"${RESTART_DIR}\", njob=\"361\", verif_case=\"${VERIF_CASE}\", verif_type=\"${VERIF_TYPE}\", vx_mask=\"${NEST}\", met_tool=\"point_stat\", vdate=\"${VDATE}\", vhour=\"${VHOUR}\", fhr_start=\"${FHR_START}\", fhr_end=\"${FHR_END}\", fhr_incr=\"${FHR_INCR}\", model=\"${MODELNAME}\", var_name=\"${VAR_NAME}\")"
        python -c 'import mesoscale_util; mesoscale_util.mark_job_completed("/lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_mesoscale_nam_grid2obs_stats_00.178207783.cbqs01/grid2obs/completed_jobs.txt", "job361", job_type="generate")'
        if [ -f /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_mesoscale_nam_grid2obs_stats_00.178207783.cbqs01/grid2obs/completed_jobs.txt ] && [ $SENDCOM == YES ]; then cp -rpfv /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_mesoscale_nam_grid2obs_stats_00.178207783.cbqs01/grid2obs/completed_jobs.txt /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr652/evs/v2.0/stats/mesoscale/atmos.20250126/nam/grid2obs/restart/c07; fi              
```

Here's where those commented out lines are useful.  It shows what has already been run in the previous interrupted run.

But you see, we get to the first line without a comment (#) in front of it is the call to mesoscale_stats_grid2obs_created_merged_ptype.py.  I don't understand why it hits there, but I understand the error since in the restart run it doesn't run RegridDataPlane.  All of this should be commented, I would think.

PerryShafran-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/mesoscale_restart

PerryShafran-NOAA commented:
@malloryprow As denoted in the Discussion item, I think the problem is that the indentations in the mesoscale_stats_grid2obs_create_job_script.py code were not correct, thus having some items in the ptype section outside the if block that operates on this code.  That's why some of the items in job361 were uncommented out when they should have been commented.  I am doing some testing to confirm this.

PerryShafran-NOAA commented:
Ah, yes, indeed, that was the reason for the error.  Latest tests worked well - no warnings or errors in the restart run!

Commit:
Updating indentation in script to fix errors

PerryShafran-NOAA commented:
Updates are made.  You should do another restart test with the updated code.

malloryprow commented:
COMOUT  is /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr652/evs/v2.0/stats/mesoscale.

### 1. jevs_mesoscale_nam_grid2obs_stats.sh (vhr=07) - Restart Test
**Full Test Run Runtime**: 19 minutes
**Interrupted Runtime**: 15 minutes
**Interrupted Log File**: /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr652/EVS/dev/drivers/scripts/stats/mesoscale/jevs_mesoscale_nam_grid2obs_stats_00.o178553723
**Interrupted DATA**: /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_mesoscale_nam_grid2obs_stats_00.178553723.cbqs01
**Restart Runtime**: 19 minutes
**Restart Log File**: /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr652/EVS/dev/drivers/scripts/stats/mesoscale/jevs_mesoscale_nam_grid2obs_stats_00.o178556499
**Restart DATA**: /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_mesoscale_nam_grid2obs_stats_00.178556499.cbqs01

### 2. jevs_mesoscale_rap_grid2obs_stats.sh (vhr=07) - Restart Test
**Full Test Run Runtime**: 26 minutes
**Interrupted Runtime**: 15 minutes
**Interrupted Log File**: /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr652/EVS/dev/drivers/scripts/stats/mesoscale/jevs_mesoscale_rap_grid2obs_stats_00.o178553725
**Interrupted DATA**: /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_mesoscale_rap_grid2obs_stats_00.178553725.cbqs01
**Restart Run Time**: 26 minutes
**Restart Log File**: /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr652/EVS/dev/drivers/scripts/stats/mesoscale/jevs_mesoscale_rap_grid2obs_stats_00.o178556549
**Restart DATA**: /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_mesoscale_rap_grid2obs_stats_00.178556549.cbqs01

PerryShafran-NOAA commented:
Finally, everything is good!  No errors/warnings/not founds, final stat files match the full run stat files.  I think everything here is perfect.

malloryprow reviewed (approved):
Code changes are good and testing successful. 

There are a few concerns I have but I also have them for cam so I can send an email about them and hopefully a solution can be worked together for both components.

Thanks @PerryShafran-NOAA! Cross this off the Fixes and Additions document.

AndrewBenjamin-NOAA reviewed (approved):
I have reviewed these changes and approved this PR. Thanks to everyone involved for their input and diligence in getting this PR ready.
